prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
+		p.Strategy = strategy
 	}
+}
 
-	return false
+// WithNamespace can be used to set the Namespace field when configuring a Pruner
+func WithNamespace(namespace string) PrunerOption {
+	return func(p *Pruner) {
+		p.Namespace = namespace
+	}
 }
 
-// containsName checks if a string is present in a ResourceInfo slice
-func containsName(s []ResourceInfo, str string) bool {
-	for _, v := range s {
-		if v.Name == str {
-			return true
-		}
+// WithLogger can be used to set the Logger field when configuring a Pruner
+func WithLogger(logger logr.Logger) PrunerOption {
+	return func(p *Pruner) {
+		p.Logger = logger
 	}
+}
 
-	return false
+// WithLabels can be used to set the Labels field when configuring a Pruner
+func WithLabels(labels map[string]string) PrunerOption {
+	return func(p *Pruner) {
+		p.Labels = labels
+	}
 }
-func (config Config) validate() (err error) {
 
-	if config.CustomStrategy == nil && config.Strategy.Mode == CustomStrategy {
-		return fmt.Errorf("custom strategies require a strategy function to be specified")
+// WithGVK can be used to set the GVK field when configuring a Pruner
+func WithGVK(gvk schema.GroupVersionKind) PrunerOption {
+	return func(p *Pruner) {
+		p.GVK = gvk
 	}
+}
 
-	if len(config.Namespaces) == 0 {
-		return fmt.Errorf("namespaces are required")
+// NewPruner returns a pruner that uses the given strategy to prune objects.
+func NewPruner(prunerClient client.Client, opts ...PrunerOption) Pruner {
+	pruner := Pruner{
+		Registry: defaultRegistry,
+		Client:   prunerClient,
+		Logger:   Logger(context.Background(), Pruner{}),
+		GVK:      corev1.SchemeGroupVersion.WithKind("Pod"),
 	}
 
-	if containsString(config.Namespaces, "") {
-		return fmt.Errorf("empty namespace value not supported")
+	for _, opt := range opts {
+		opt(&pruner)
 	}
 
-	_, err = labels.Parse(config.LabelSelector)
-	if err != nil {
-		return err
+	return pruner
+}
+
+// Prune runs the pruner.
+func (p Pruner) Prune(ctx context.Context) ([]client.Object, error) {
+	var objs []client.Object
+	p.Logger.Info("Starting the pruning process...")
+	listOpts := client.ListOptions{
+		LabelSelector: labels.Set(p.Labels).AsSelector(),
+		Namespace:     p.Namespace,
+	}
+
+	var unstructuredObjs unstructured.UnstructuredList
+	unstructuredObjs.SetGroupVersionKind(p.GVK)
+	if err := p.Client.List(ctx, &unstructuredObjs, &listOpts); err != nil {
+		return nil, fmt.Errorf("error getting a list of resources: %w", err)
 	}
 
-	if config.Strategy.Mode == MaxAgeStrategy {
-		_, err = time.ParseDuration(config.Strategy.MaxAgeSetting)
+	for _, unsObj := range unstructuredObjs.Items {
+		obj, err := convert(p.Client, p.GVK, &unsObj)
 		if err != nil {
-			return err
+			return nil, err
+		}
+
+		if err := p.IsPrunable(obj); isUnprunable(err) {
+			p.Logger.Info(fmt.Errorf("object is unprunable: %w", err).Error())
+			continue
+		} else if err != nil {
+			return nil, err
 		}
+
+		objs = append(objs, obj)
+	}
+
+	objsToPrune, err := p.Strategy(ctx, objs)
```

Comment:
I haven't seen Strategy being defaulted like the other fields. It may not been guaranteed at compile time that it is not null as it is not an explicit parameter of NewPruner either

Reply:
Thanks for bringing this up as well. Similar to my response above, Ryan and I spoke about this offline yesterday and we are adding the Strategy to the explicit parameters of NewPruner

Reply:
done in https://github.com/operator-framework/operator-lib/pull/105/commits/67688741005b777c097f0c3952c3d6c010d15db9